[Dependency Cache] Dependency Cache on CI per Project [with GRADLE_RO_DEP_CACHE]

ParaskP7 opened the pull request:
Related: #135

This change mainly adds a project prefix (`BUILDKITE_PIPELINE_SLUG`) to the existing gradle dependency cache key (`GRADLE_DEPENDENCY_CACHE_KEY`) so that to enable testing the existing dependency cache mechanism, but per project, while depending on the read-only cache ([GRADLE_RO_DEP_CACHE](https://docs.gradle.org/current/userguide/dependency_caching.html#sec:shared-readonly-cache)).

FYI: The rest of the optimization are done for measuring purposes (like adding the `duration`, `cache size`, etc).

---

- [x] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary.

ParaskP7 committed:
Gradle: Add project prefix to gradle dependency cache key

dangermattic commented:
<!--
  0 Errors
  1 Warning: Please add an entry in the <co...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please add an entry in the <code>CHANGELOG.md</code> file to describe the changes made by this PR</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

ParaskP7 committed:
Gradle: Echo on gradle ro dep cache environment variable

FYI: This change will help us identify if this 'GRADLE_RO_DEP_CACHE' is
set or not, and thus, if the Gradle shared read-only dependency cache
feature is enabled or not.

ParaskP7 committed:
Gradle: Replace project prefix with the standard buildkite pipeline slug

This is done in order to avoid having each client project adding its own
project prefix to distinguish the Gradle dependency cache key.

Test Only PR (WCAndroid): https://github.com/woocommerce/
woocommerce-android/pull/13122

For example, for WCAndroid, this would mean that instead of using this
'WOOCOMMERCE' value via this 'GRADLE_DEPENDENCY_CACHE_PROJECT_PREFIX'
environment variable, which ultimately is creating the below
'WOOCOMMERCE_GRADLE_DEPENDENCY_CACHE' key, instead, with this change the
'woocommerce-android_GRADLE_DEPENDENCY_CACHE' key will be
automatically created, taking the 'woocommerce-android' value from the
standard 'BUILDKITE_PIPELINE_SLUG' environment variable.

For more info see: p1734350857095859-slack-C020Q0Z579V

ParaskP7 committed:
Gradle: Remove echo on gradle ro dep cache environment variable

Previously, and based on these #25 and #26 PRs, it was incorrectly
assumed that this 'GRADLE_RO_DEP_CACHE' environment variable is being
driven by the '.mobile-secrets' repo. However, after trial-and-error it
became evident that the 'buildkite-ci' repo is driving this via this #72
change that adds the 'GRADLE_RO_DEP_CACHE' environment variable into the
'Provision Android Image' via the 'android-playbook.yml'. This means
that this 'GRADLE_RO_DEP_CACHE' environment variable will always be set
no matter what, and will be always pointing to:

buildkite_agent_home + gradle_home + gradle_readonly_dependency_cache
/var/lib/buildkite-agent + /.gradle + /read-only-dependency-cache

PRs (a8c-ci-toolkit-buildkite-plugin):
- Adds save & restore gradle dependency cache scripts #25
https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/25
- Use dirname & basename for Gradle dependency cache logic #26
https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/26

PR (buildkite-ci)
- Add Google Cloud to Amazon AMI #72 -> An improvement to support Gradle
Caching.
https://github.com/Automattic/buildkite-ci/pull/72

ParaskP7 committed:
Gradle: Measure how long it takes to save/restore the cache

ParaskP7 committed:
Gradle: Measure the (un)compressed cache size during save/restore

ParaskP7 committed:
Gradle: Move cache duration measurement on a separate line

ParaskP7 committed:
Gradle: Convert reported cache size from bytes to megabytes

ParaskP7 committed:
Gradle: Add optional cache file to restore cache script

ParaskP7 commented:
Related: #135
